flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/TypeStrategies.java

```diff
@@ -105,6 +105,23 @@ public static TypeStrategy nullableIfArgs(TypeStrategy initialStrategy) {
         return nullableIfArgs(ConstantArgumentCount.any(), initialStrategy);
     }
 
+    /**
+     * A type strategy that can be used to make a result type nullable if all the selected input
+     * arguments is nullable. Otherwise the type will be not null.
+     */
+    public static TypeStrategy nullableIfAllArgs(
```

Comment:
given that the `COMMON_ARG` input strategy is executed first, can't we just derive the nullability from the first arg?

Reply:
Deriving nullability from the first arg will bring back problems like:

```
COALESCE(REGEXP_EXTRACT('22','[A-Z]+'),'-')
```

Returning a nullable data type, because `REGEXP_EXTRACT` is always nullable. Is that what you mean?

Reply:
Yeah, I don't think it would work either. The common argument strategy will make the type nullable if any type is nullable, but we want nullability when _all_ are nullable.

Reply:
While looking in the calcite doc, I've found this which confirms the correctness of our type inference logic: https://calcite.apache.org/docs/reference.html#conversion-contexts-and-strategies